src/Interpreters/InterpreterAlterQuery.cpp

```diff
@@ -173,37 +176,9 @@ BlockIO InterpreterAlterQuery::executeToTable(const ASTAlterQuery & alter)
             /// ALTER TABLE ... DELETE WHERE 1 should be executed as truncate
             if (mut_command->type == MutationCommand::DELETE && isAlwaysTruePredicate(mut_command->predicate))
             {
-                auto context = getContext();
-                context->checkAccess(AccessType::TRUNCATE, table_id);
-                table->checkTableCanBeDropped(context);
-
-                auto metadata_snapshot = table->getInMemoryMetadataPtr();
-
-                TableExclusiveLockHolder table_excl_lock;
-                /// We don't need any lock for ReplicatedMergeTree and for simple MergeTree
-                /// For the rest of tables types exclusive lock is needed
-                if (!std::dynamic_pointer_cast<MergeTreeData>(table))
-                    table_excl_lock = table->lockExclusively(context->getCurrentQueryId(), context->getSettingsRef()[Setting::lock_acquire_timeout]);
-
-                String truncate_query = "TRUNCATE TABLE " + table->getStorageID().getFullTableName()
-                    + (alter.cluster.empty() ? "" : " ON CLUSTER " + backQuoteIfNeed(alter.cluster));
-
-                ParserDropQuery parser;
-                auto current_query_ptr = parseQuery(
-                    parser,
-                    truncate_query.data(),
-                    truncate_query.data() + truncate_query.size(),
-                    "ALTER query",
-                    0,
-                    DBMS_DEFAULT_MAX_PARSER_DEPTH,
-                    DBMS_DEFAULT_MAX_PARSER_BACKTRACKS);
-
-                /// Drop table data, don't touch metadata
-                table->truncate(current_query_ptr, metadata_snapshot, context, table_excl_lock);
-                return {};
+                is_truncate = true;
```

Comment:
We should also make sure there are no other commands,  just set it back to `false` on any other command and "non-truncate" mutations. Yes, optimization will be omitted in that case, but IMO it should not be a big deal.

Reply:
I don't think there is a case where we have truncate and multiple commands, but I added it for safety. Also it is not enough to just set `is_truncate=false` because the truncate command might be the last one, in which case we will reject our undo, so I added a new variable to check for optimization undo.